eval: Make writefile() able to disable fsync()

ZyX-I opened the pull request:
Ref #6420

ZyX-I committed:
eval: Make writefile() able to disable fsync()

ZyX-I commented:
As fsync() is a safer option, it is on by default. As other flags enable something and are lowercase, flag that disables fsync() is uppercase.

justinmk commented:
Shouldn't writefile() default behavior be decided by 'fsync' option? And then overridden with s/S flag pair.

blueyed commented:
@justinmk 
Good idea, but the documentation for `'fsync'` could be interpreted as not applying here (but only to `:w` etc).
I'm assuming that Vim does not `fsync` with `writefile()`, does it?

blueyed commented:
Just for reference: any flag for `writefile` will trigger the `binary` flag in old Vim versions, where a trueish value enables this, and no other flags were known yet.  So using `S` would always trigger it, but that should be handled in plugins by checking for a patch level probably.

justinmk commented:
> Good idea, but the documentation for 'fsync' could be interpreted as not applying here (but only to :w etc).

Right, but I'm suggesting that we apply it anyway.

ZyX-I commented:
@blueyed Vim does not `fsync()` neither with `writefile()` nor when writing viminfo (second user of buffered writing is ShaDa). Basically you can’t use `FILE*`, `fsync` and keep up with the standard: even if `fileno()` is available (it is POSIX, not C99), it is required that you should not mix unistd API with stdio API, this may yield unexpected results.

Though `fsync()` with the whole fd-related API is also BSD/POSIX/SUS and not C99, yet available on Windows (but specifically `fsync()` is absent, libuv uses `FlushFileBuffers` and we are using libuv).

ZyX-I commented:
`ag` finds exactly three occurrences of `p_fs` (fsync option): fileio.c with `#ifdef UNIX` guard (BTW, I think I should remove it, libuv has windows implementation for that), globals.h and options.lua.

Note that our help does not warn about “systems without fsync implementation” like Vim documentation. Yet there is a guard around the only place where `p_fs` is checked.

justinmk commented:
> ag finds exactly three occurrences of p_fs (fsync option): fileio.c with #ifdef UNIX guard (BTW, I think I should remove it, libuv has windows implementation for that), 

Yes, probably should remove it. Most likely we overlooked it because Windows was "todo" for a long time. But the documentation was updated in anticipation of convergence.

ZyX-I committed:
fileio: Refactor errmsg handling

Adds os_strerror() result to a number of places. Also since I could not track 
where err\* variables are NULL and where they are not, using macros to make sure 
that all three variables are set at once.

Removes #ifdef UNIX around the use of os_fsync, makes it use os_close in place 
of close in some places.

ZyX-I committed:
message: Do not use IObuff in emsgf

ZyX-I committed:
fixup!

ZyX-I committed:
fileio: Clean up IObuff-manipulation mess

ZyX-I committed:
fileio: Refactor msg_add_fname to something which needs no comments

ZyX-I committed:
eval: Add s flag, use p_fs by default, error out on unknown flag

ZyX-I committed:
fileio: Save details about E212 error

ZyX-I committed:
functests: Test some :write errors

ZyX-I committed:
tests: Fix testlint errors

ZyX-I commented:
My watchdog is working (which restarts single includes test), but this is not very useful: “hang up” was replaced with “CMakeFiles/test-includes-api-window-h.dir/auto/api/window.h.test-include.c.o: file not recognized: File truncated” and “/usr/bin/ld: cannot find CMakeFiles/test-includes-mark_defs-h.dir/auto/mark_defs.h.test-include.c.o: No such file or directory”.

ZyX-I commented on test/functional/ui/cursor_spec.lua:
```diff
@@ -1,8 +1,7 @@
-local insert, execute = helpers.insert, helpers.execute
-local eq, funcs = helpers.eq, helpers.funcs
+local clear, meths = helpers.clear, helpers.meths
+local eq = helpers.eq
```
BTW, how that appeared in master? I did not touch the file, yet tests fail.

justinmk replied:
my fault

ZyX-I committed:
fileio: Fix most linter errors

One error is still kept: buf_write function is too large.

ZyX-I committed:
ci: Do not mark test as failed if it is previous one which failed

ZyX-I commented:
Will have to merge despite linter error: I am not really up to splitting `buf_write`.

blueyed commented:
I wonder if it should really use fsync by default?!
Typically you call this manually, or implicit when closing a file.
(A file cannot be closed though in VimL, can it?)

ZyX-I commented:
@blueyed Typically software does not bother to call `fsync()`, it is never done implicitly and “manually” you can only fsync all files in the whole system. What does call fsync() is something old-school like Vim: software whose developers assume that crashing system is relatively common. E.g. in Python you will get `fsync()` only if you use `posix` module, neither `file.flush()` nor `file.close()` do that.

ZyX-I commented:
@blueyed And `writefile()` does close file. As well as e.g. `:write` first opens it and then closes back (after writing).

ZyX-I commented:
When file is opened in a buffer it is not opened from the system POV: file descriptor is created and associated with the file only temporary, when you read or write a file, disappearing after finishing reading or writing.

justinmk commented:
> I wonder if it should really use fsync by default?!

You can control the default, e.g. `:set nofsync`.
